boards: arm: add QEMU support for Cortex-M0

ioannisg opened the pull request:
QEMU for Cortex-M0, based on BBC Microbit

Fixes #2686

ioannisg committed:
boards: arm: add QEMU support for Cortex-M0

This commit defines qemu_cortex_m0 board, adding
support for Cortex-M0 in QEMU. The added platform
is based on the (nRF51) bbc_microbit board.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: qemu_cortex_m0: adding documentation

This commit adds some documentation for the
newly introduced qemu_cortex_m0 platform.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
CODEOWNERS: adding code owner for ARM Qemu Cortex-M* platforms

Adding code owner for ARM boards
- qemu_cortex_m0
- qemu_cortex_m3

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

galak reviewed (commented):
can we not just use the microbit board?  do we need to introduce a different board port?

nashif commented:
> can we not just use the microbit board? do we need to introduce a different board port?

agree here, similar to what we did with `mps2_an385`

nashif commented:
also, I would make this a default platform and remove the nrf51 board which was covering for cortex-m0 as a default platform.

nashif commented:
> > can we not just use the microbit board? do we need to introduce a different board port?
> 
> agree here, similar to what we did with `mps2_an385`

although, once issue we have with microbit is the low RAM which makes it less interesting when it comes to coverage...

ioannisg commented:
> > > can we not just use the microbit board? do we need to introduce a different board port?
> > 
> > 
> > agree here, similar to what we did with `mps2_an385`
> 
> although, one issue we have with microbit is the low RAM which makes it less interesting when it comes to coverage...

Yes, and besides that, @nashif @galak :+1: 
My goal was to have a generic Cortex-M0 platform to get kernel/arch coverage
BBC Microbit is an actual IoT platform used by lots of developers, so I don't want to confuse with adding qemu support on that one. Also bbc microbit uses peripheral devices and has additional hardware that I had no intention of testing via the QEMU platform.

that's why I was happy to add an additional platform, similarly to qemu_cortex_m3.

ioannisg committed:
boards: arm: qemu_cortex_m0: configure board with own timer driver

This commit configures the qemu_cortex_m0 board to
build with it's custom timer driver, instead of the
default nrf_rtc_timer driver for nRF51x SoCs. It,
additionally, configures a default system clock
frequency to 1MHz, as well as 10 Hz tick frequency.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
drivers: timer: nrf: minor comment fix

Fix an inline comment in nrf_rtc_timer.c correcting the
path to the mentioned test.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: qemu_cortex_m0: implement custom system clock driver

For the qemu_cortex_m0 we implement a custom system clock
driver based on the nRF51 TIMER peripheral. The system
clock is configured to run at 1 MHz frequency.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: qemu_cortex_m0: filter out a couple of failing tests

We filter out the following kernel tests
- tickless_concept
- timer_api
from the set of tests running on QEMU Cortex-M0 platform,
as the tests consistently fail on QEMU. In addition, we
add a workaround for kernel/interrupt test, so it can
successfully execute on QEMU Cortex-M0.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
I added a timer driver for the qemu_cortex_m0 platform, based on the nRF TIMER peripheral. The driver looks very similar to  nrf_rtc_timer driver in drivers/timer. With this timer driver, the emulated microbit platform is able to generate timer interrupts, and effectively, run almost all the kernel test suite successfully.

As our goal is to get coverage on kernel and arm code on Cortex-M0, I believe we can live with the filtering-out the failing tests for this platform.

ioannisg commented on boards/arm/qemu_cortex_m0/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+This configuration provides support for an ARM Cortex-M0 CPU and these devices:
+
+* Nested Vectored Interrupt Controller
+* RTC (nRF RTC System Clock)
```
Needs to be changed to TIMER

ioannisg commented on boards/arm/qemu_cortex_m0/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+System Clock
+------------
+
+This board configuration uses a system clock frequency of 32768 Hz.
```
1 MHz

nashif commented on boards/arm/qemu_cortex_m0/qemu_cortex_m0.yaml:
```diff
@@ -0,0 +1,14 @@
+  - xtools
+ram: 16
+flash: 256
+testing:
```
need to make this a default board.

ioannisg replied:
yeap

ioannisg replied:
done @nashif

andyross commented on boards/arm/qemu_cortex_m0/nrf_timer_timer.c:
```diff
@@ -0,0 +1,188 @@
+/*
```
Why did the whole driver have to be copied to the board directory?

ioannisg replied:
That's because there is no point having a second system clock driver based on Timer peripheral , besides the bed RTF driver. I only added the driver because the qemu does not emulate the RTC  :)

andyross replied:
Sorry, I'm confused. Where does the emulation come from then? I mean, it looks like nrf_timer_cc_read() is returning your counter value from somewhere, right? So the HAL is doing the emulation? If that's true, why won't the default driver work?

ioannisg replied:
@andyross the default nRF driver, used in Zephyr, is based on the NRF RTC peripheral. This peripheral is, however, not supported in the QEMU emulation of the Microbit platform (nRF51). Instead, the QEMU guys have added support for the NRF TIMER peripheral. So, I had to write a system timer driver, based on the NRF TIMER peripheral.

But there is no need to have this driver in the /drivers/timer directory as a generic timer driver; cause there is no use case for that; nRF SoCs use the nRF RTC timer that can utilize the low-frequency low-power clock source. This driver is not, even, specific to Microbit; microbit can use the default nRF RTC driver. This driver is specific for the qemu_cortex_m0 platform. Its sole purpose is to add system timer support for qemu cortex-m0, so we get coverage on ARMv6-M architecture.

ioannisg replied:
> nrf_timer_cc_read() is returning your counter value from somewhere, right? So the HAL is doing the emulation? If that's true, why won't the default driver work?

Correct. The default driver is a different peripheral. The "corresponding function, there, is called `nrf_rtc_cc_read()`. But as I said, this peripheral does not _yet_ work in qemu.

ioannisg commented:
> Can we give this new target significantly more RAM & flash so we can run all tests? From browsing this PR seems like we only have 16K of RAM.

I'll check this, @andrewboie

ioannisg commented:
> > Can we give this new target significantly more RAM & flash so we can run all tests? From browsing this PR seems like we only have 16K of RAM.
> 
> I'll check this, @andrewboie

@andrewboie I have looked into modifying DTS for the board, so it adds more ram and flash, however, several of the additional tests that are now attempted, are resulting in QEMU core dump; perhaps QEMU forbids using non-existing memory areas on this particular board. So, I would leave this out for now, to get basic cortex-M0 qemu support and look at it again in the future

nashif reviewed (approved):
Should use this PR to remove nrf52_pca10028 as default, it was representing the cortex-m0 family:

```
--- a/boards/arm/nrf51_pca10028/nrf51_pca10028.yaml
+++ b/boards/arm/nrf51_pca10028/nrf51_pca10028.yaml
@@ -15,6 +15,5 @@ supported:
   - watchdog
   - counter
 testing:
-  default: true
   ignore_tags:
     - net
```

ioannisg commented:
> is PR to remove nrf52_pca10028 as default, it was represe

Alright, I can send a patch upstream after this PR is merged.

ioannisg commented:
FYI: we currently build/run 104 tests for qemu_cortex_m0 - this looks like good coverage, IMO ;)

ioannisg commented:
@andyross @galak @andrewboie could you re-review?
@dbkinder could you check the docs here?
